### PR TITLE
Paragraph - enable paragraph to have a custom content

### PR DESCRIPTION
**Pull request by jgonza16**

### CODE REVIEW

**Comment by jgonza16** [comment]
hello @daniele-zurico, changes is done. I have made value take precedence over children because it is the previous property, how do you see it?

File: src/paragraph/Paragraph.tsx [hunk 1]
```diff
@@ -7,21 +7,33 @@ type ParagraphProps = {
...
+  value,
+  props,
+  children,
+}: ParagraphProps): JSX.Element => {
   const dynamicClassName = classNames(['dcx-paragraph', className]);
-  
+  let content: typeof value | typeof children;
+  if (children) content = children;
+  if (value) content = value;
+
```
Comment (daniele-zurico) [inline src/paragraph/Paragraph.tsx:33]:
```suggestion
  if (value === undefined && children === undefined) {
    throw new Error(
      'Paragraph requires a value or a custom content is defined'
    );
  }
```

File: src/paragraph/__tests__/Paragraph.test.tsx [hunk 2]
```diff
@@ -4,22 +4,36 @@ import { render, screen } from '@testing-library/react';
...
+  it('should be value prop more priority then custom content', () => {
+    const valueText = 'a custom value';
+    const customText = 'a custom content';
+    render(
+      <Paragraph className="paragraph" value={valueText}>
+        {customText}
+      </Paragraph>
+    );
+    expect(screen.getByText(valueText)).toBeInTheDocument();
+  });
```
Comment (daniele-zurico) [inline src/paragraph/__tests__/Paragraph.test.tsx:30]: Something like that:
```
it('should throw an error if content or children are not passed, () => {
    expect(render(<Paragraph className="paragraph"/>)).toThrow('Paragraph requires that a value or a custom content is defined');
});
```

**Comment by daniele-zurico** [comment]
> hello @daniele-zurico, changes is done. I have made value take precedence over children because it is the previous property, how do you see it?

hey @jgonza16 first thing first thanks a lot for your speedy PR it is really solid. I just left few comments most of them can be directly committed without you to copy the content of it.
The main observation I put in your PR is the fact that at least one between content and custom content need to be present in the Paragraph component for two reasons:
1. before value was mandatory so it was expected to have a value whilst now we moved to optional
2. a Paragraph without content os not a paragraph.
Let me know if you agree on that

File: src/paragraph/__tests__/Paragraph.test.tsx (hunk 2 above)
Comment (jgonza16) [reply src/paragraph/__tests__/Paragraph.test.tsx:30]: with typescript validate types we can improve the development experience and remove this test

**Comment by jgonza16** [comment]
@daniele-zurico yes you are right. we can use typescript better for this. we pushed changes in paraagraph.tsx. this is i approach:
```tsx
type Props = {
  /**
   * optional CSS class name
   */
  className?: string;
  /**
   * content of the paragraph. it has precedence to children
   */
  value: string | number;
  /**
   * allow to specify a custom content
   */
  children: string | number | JSX.Element;
  /**
   * Additional props/attributes
   */
  props?: React.HTMLAttributes<HTMLParagraphElement>;
};

type ParagraphProps = Omit<Props, 'value'> | Omit<Props, 'children'>;
type Value = Pick<Props, 'value'>;
type Children = Pick<Props, 'children'>;

const isValueType = (t: any): t is Value => !!t.value;
const isChildrenType = (t: any): t is Children => !!t.children;
```
into the compoent check: 
```tsx
  if (isValueType(rest)) content = rest.value;
  if (isChildrenType(rest)) content = rest.children;
```

**Comment by daniele-zurico** [comment]
> @daniele-zurico yes you are right. we can use typescript better for this. we pushed changes in paraagraph.tsx. this is my approach:
> 
> ```tsx
> type Props = {
>   className?: string;
>   value: string | number;
>   children: string | number | JSX.Element;
>   props?: React.HTMLAttributes<HTMLParagraphElement>;
> };
> 
> type ParagraphValue = Omit<Props, 'children'>;
> type ParagraphChildren = Omit<Props, 'value'>;
> type ParagraphProps = ParagraphValue | ParagraphChildren;
> 
> const isValueType = (p: any): p is ParagraphValue => !!p.value;
> const isChildrenType = (p: any): p is ParagraphChildren => !!p.children;
> ```
> 
> into the compoent check:
> 
> ```tsx
> export const Paragraph = ({
>   className,
>   props,
>   ...rest
> }: ParagraphProps): JSX.Element => {
> 
>   if (isChildrenType(rest)) content = rest.children;
>   if (isValueType(rest)) content = rest.value;
> 
> };
> ```
hey @jgonza16 we can go through this approach as well (that is the original I suggested) however I didn't progress with that because I don't know how the storybook will behave with the 3 types on the following page: https://main--6069a6f47f4b9f002171f8e1.chromatic.com/?path=/docs/dcxlibrary-typography-paragraph-documentation--docs (on the table at the end of the page)
I think it will not be able to display that and will be missed in the documentation.

**Review by daniele-zurico — APPROVED** [review]

**Pull request merged**